storage/feed_query_builder.go

```diff
@@ -0,0 +1,300 @@
+func (f *FeedQueryBuilder) WithCounters() *FeedQueryBuilder {
+	f.withCounters = true
+	return f
+}
+
+// WithOrder set the sorting order.
+func (f *FeedQueryBuilder) WithOrder(order string) *FeedQueryBuilder {
+	f.order = order
+	return f
+}
+
+// WithDirection set the sorting direction.
+func (f *FeedQueryBuilder) WithDirection(direction string) *FeedQueryBuilder {
+	f.direction = direction
+	return f
+}
+
+// WithLimit set the limit.
+func (f *FeedQueryBuilder) WithLimit(limit int) *FeedQueryBuilder {
+	f.limit = limit
+	return f
+}
+
+// WithOffset set the offset.
+func (f *FeedQueryBuilder) WithOffset(offset int) *FeedQueryBuilder {
+	f.offset = offset
+	return f
+}
+
+func (f *FeedQueryBuilder) buildCondition() string {
+	return strings.Join(f.conditions, " AND ")
+}
+
+func (f *FeedQueryBuilder) buildCounterCondition() string {
+	return strings.Join(f.counterConditions, " AND ")
+}
+
+func (f *FeedQueryBuilder) buildSorting() string {
+	var parts []string
+
+	if f.order != "" {
+		parts = append(parts, fmt.Sprintf(`ORDER BY %s`, f.order))
+	}
+
+	if f.direction != "" {
+		parts = append(parts, fmt.Sprintf(`%s`, f.direction))
+	}
+
+	if f.limit != 0 {
+		parts = append(parts, fmt.Sprintf(`LIMIT %d`, f.limit))
+	}
+
+	if f.offset != 0 {
+		parts = append(parts, fmt.Sprintf(`OFFSET %d`, f.offset))
+	}
+
+	return strings.Join(parts, " ")
+}
+
+// GetFeed returns a single feed that match the condition.
+func (f *FeedQueryBuilder) GetFeed() (*model.Feed, error) {
+	f.limit = 1
+	feeds, err := f.GetFeeds()
+	if err != nil {
+		return nil, err
+	}
+
+	if len(feeds) != 1 {
+		return nil, nil
+	}
+
+	return feeds[0], nil
+}
+
+// GetFeeds returns a list of feeds that match the condition.
+func (f *FeedQueryBuilder) GetFeeds() (model.Feeds, error) {
+	var query = `
+		SELECT
+			f.id,
+			f.feed_url,
+			f.site_url,
+			f.title,
+			f.etag_header,
+			f.last_modified_header,
+			f.user_id,
+			f.checked_at at time zone u.timezone,
+			f.parsing_error_count,
+			f.parsing_error_msg,
+			f.scraper_rules,
+			f.rewrite_rules,
+			f.blocklist_rules,
+			f.keeplist_rules,
+			f.crawler,
+			f.user_agent,
+			f.username,
+			f.password,
+			f.ignore_http_cache,
+			f.fetch_via_proxy,
+			f.disabled,
+			f.category_id,
+			c.title as category_title,
+			fi.icon_id,
+			u.timezone
+		FROM
+			feeds f
+		LEFT JOIN
+			categories c ON c.id=f.category_id
+		LEFT JOIN
+			feed_icons fi ON fi.feed_id=f.id
+		LEFT JOIN
+			users u ON u.id=f.user_id
+		WHERE %s 
+		%s
+	`
+
+	query = fmt.Sprintf(query, f.buildCondition(), f.buildSorting())
+
+	rows, err := f.store.db.Query(query, f.args...)
+	if err != nil {
+		return nil, fmt.Errorf(`store: unable to fetch feeds: %w`, err)
+	}
+	defer rows.Close()
+
+	readCounters, unreadCounters, err := f.fetchFeedCounter()
+	if err != nil {
+		return nil, err
+	}
+
+	feeds := make(model.Feeds, 0)
+	for rows.Next() {
+		var feed model.Feed
+		var iconID interface{}
```

Comment:
`iconID` should be a `sql.NullInt64`

`interface{}` was wrong in initial code.